cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java

```diff
@@ -231,35 +240,46 @@ public void starting() {
           // If RunId is not time-based, use current time as start time
           startTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
         }
-        if (workflowName == null) {
+        if (workflowRunId == null) {
           store.setStart(program.getId(), runId.getId(), startTimeInSeconds, twillRunId,
                          userArgs.asMap(), arguments.asMap());
         } else {
           // Program started by Workflow
-          store.setWorkflowProgramStart(program.getId(), runId.getId(), workflowName, workflowRunId, workflowNodeId,
-                                        startTimeInSeconds, twillRunId);
+          store.setWorkflowProgramStart(program.getId(), runId.getId(), workflowRunId.getProgram(),
+                                        workflowRunId.getRun(), workflowNodeId, startTimeInSeconds, twillRunId);
         }
       }
 
       @Override
       public void terminated(Service.State from) {
         closeAllQuietly(closeables);
+        ProgramRunStatus runStatus = ProgramController.State.COMPLETED.getRunStatus();
         if (from == Service.State.STOPPING) {
           // Service was killed
+          runStatus = ProgramController.State.KILLED.getRunStatus();
+        }
+
+        if (workflowRunId == null) {
```

Comment:
This logic is pretty brutal. Is there a way for the Store.setStop to figure out if a given `ProgramRunId` has a workflow as parent and just do the related RunRecords update as well? Ideally the same should apply for the setStart as well.